File: go/adapters/compute_wrap_test.go
// Serverless handler wraps for compute capture.
// Mirrors python/tests/test_compute_wrap.py.

package adapters

import (
	"context"
	"errors"
	"testing"

	"github.com/google/uuid"

	"github.com/DexwoxBusiness/dexcost-go/core"
)

// computeEventCollector is a minimal Buffer implementation that captures
// inserted events in memory — used by the wrap tests to verify emission
// without spinning up the full SQLite tracker.
type computeEventCollector struct {
	events []core.Event
}

func (b *computeEventCollector) InsertTask(t core.Task) error    { return nil }
func (b *computeEventCollector) UpdateTask(t core.Task) error    { return nil }
func (b *computeEventCollector) GetTask(id string) (*core.Task, error) {
	return nil, nil
}
func (b *computeEventCollector) InsertEvent(e core.Event) error {
	b.events = append(b.events, e)
	return nil
}
func (b *computeEventCollector) UpdateEvent(e core.Event) error  { return nil }
func (b *computeEventCollector) QueryEvents(taskID string) ([]core.Event, error) {
	return b.events, nil
}
func (b *computeEventCollector) Close() error { return nil }

func newComputeWrapTestEnv(t *testing.T) (*computeEventCollector, *core.Task) {
	t.Helper()
	core.ResetComputeRegistryForTests()
	t.Cleanup(core.ResetComputeRegistryForTests)
	buf := &computeEventCollector{}
	SetComputeEventBuffer(buf)
	t.Cleanup(func() { SetComputeEventBuffer(nil) })
	task := &core.Task{TaskID: uuid.New()}
	return buf, task
}

func TestLambdaWrapEmitsEventWithCostPending(t *testing.T) {
	buf, task := newComputeWrapTestEnv(t)
	t.Setenv("AWS_LAMBDA_FUNCTION_MEMORY_SIZE", "512")
	t.Setenv("AWS_LAMBDA_INITIALIZATION_TYPE", "on-demand")
	t.Setenv("AWS_REGION", "us-east-1")

	ctx := core.WithTask(context.Background(), task)

	captured := false
	wrapped := WrapLambdaHandler(func(ctx context.Context, evt map[string]any) (string, error) {
		captured = true
		return "ok", nil
	})
	if _, err := wrapped(ctx, map[string]any{"hello": "world"}); err != nil {
		t.Fatalf("wrapped handler err: %v", err)
	}
	if !captured {
		t.Fatal("inner handler never called")
	}
	if len(buf.events) != 1 {
		t.Fatalf("expected 1 event, got %d", len(buf.events))
	}
	ev := buf.events[0]
	if ev.EventType != core.EventTypeComputeCost {
		t.Fatalf("event type = %s, want compute_cost", ev.EventType)
	}
	if ev.Details["billing_model"] != "lambda" {
		t.Fatalf("billing_model = %v", ev.Details["billing_model"])
	}
	if ev.Details["region"] != "us-east-1" {
		t.Fatalf("region = %v", ev.Details["region"])
	}
	if pending, _ := ev.Details["cost_pending"].(bool); !pending {
		t.Fatal("cost_pending should be true at emit time")
	}
	if ev.Details["initialization_type"] != "on-demand" {
		t.Fatalf("initialization_type = %v", ev.Details["initialization_type"])
	}
}

func TestLambdaWrapPassThroughWhenNoTask(t *testing.T) {
	buf, _ := newComputeWrapTestEnv(t)

	called := false
	wrapped := WrapLambdaHandler(func(ctx context.Context, evt int) (int, error) {
		called = true
		return evt * 2, nil
	})
	out, err := wrapped(context.Background(), 21)
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if out != 42 {
		t.Fatalf("out = %d, want 42", out)
	}
	if !called {
		t.Fatal("inner not called")
	}
	if len(buf.events) != 0 {
		t.Fatalf("expected 0 events (no task in ctx), got %d", len(buf.events))
	}
}

func TestLambdaWrapHandlerErrorStillEmits(t *testing.T) {
	buf, task := newComputeWrapTestEnv(t)
	t.Setenv("AWS_LAMBDA_FUNCTION_MEMORY_SIZE", "128")
	t.Setenv("AWS_REGION", "us-east-1")

	ctx := core.WithTask(context.Background(), task)
	boom := errors.New("handler explosion")
	wrapped := WrapLambdaHandler(func(ctx context.Context, evt int) (int, error) {
		return 0, boom
	})
	_, err := wrapped(ctx, 1)
	if !errors.Is(err, boom) {
		t.Fatalf("err = %v, want %v (re-raised after capture)", err, boom)
	}
	if len(buf.events) != 1 {
		t.Fatalf("expected 1 event even on handler error, got %d", len(buf.events))
	}
}

func TestAllFiveWrapsExportedAndCallable(t *testing.T) {
	t.Setenv("AWS_LAMBDA_FUNCTION_MEMORY_SIZE", "128")
	t.Setenv("AWS_REGION", "us-east-1")

	// Compile-time check that all five wraps exist with the expected
	// signature.
	_ = WrapLambdaHandler(func(ctx context.Context, e int) (int, error) { return 0, nil })
	_ = WrapCloudRunHandler(func(ctx context.Context, e int) (int, error) { return 0, nil })
	_ = WrapCloudFunctionsHandler(func(ctx context.Context, e int) (int, error) { return 0, nil })
	_ = WrapAzureFunctionsHandler(func(ctx context.Context, e int) (int, error) { return 0, nil })
	_ = WrapVercelHandler(func(ctx context.Context, e int) (int, error) { return 0, nil })
}

File: typescript/tests/compute-wrap.test.ts
/**
 * Handler wraps emit compute_cost events per invocation and pass through
 * when no dexcost task is in context (capture spec §6 case 2).
 *
 * Ports python/tests/test_compute_wrap.py (6 cases) to vitest.
 */

import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

import * as dexcost from "../src/index.js";
import { CostTracker } from "../src/core/tracker.js";
import { wrapLambdaHandler } from "../src/adapters/compute-wrap.js";
import { runWithTask } from "../src/core/context.js";
import * as cgroup from "../src/core/cgroup-reader.js";

let tmpDir: string;
let snapshot: Record<string, string | undefined> = {};

const ENV_KEYS = [
  "AWS_LAMBDA_FUNCTION_NAME",
  "AWS_LAMBDA_FUNCTION_MEMORY_SIZE",
  "AWS_LAMBDA_INITIALIZATION_TYPE",
  "AWS_REGION",
  "AWS_DEFAULT_REGION",
];

beforeEach(() => {
  tmpDir = mkdtempSync(join(tmpdir(), "dexcost-cw-"));
  snapshot = {};
  for (const k of ENV_KEYS) {
    snapshot[k] = process.env[k];
    delete process.env[k];
  }
  // Reset the global tracker singleton if present.
  try {
    dexcost.close();
  } catch {
    // not initialized
  }
});

afterEach(() => {
  try {
    dexcost.close();
  } catch {
    // already closed
  }
  vi.restoreAllMocks();
  for (const [k, v] of Object.entries(snapshot)) {
    if (v === undefined) delete process.env[k];
    else process.env[k] = v;
  }
  rmSync(tmpDir, { recursive: true, force: true });
});

describe("wrapLambdaHandler", () => {
  test("emits a compute_cost event with cost_pending=true", async () => {
    process.env.AWS_LAMBDA_FUNCTION_NAME = "fn";
    process.env.AWS_LAMBDA_FUNCTION_MEMORY_SIZE = "1024";
    process.env.AWS_LAMBDA_INITIALIZATION_TYPE = "on-demand";
    process.env.AWS_REGION = "us-east-1";

    const tracker = dexcost.init({
      dbPath: join(tmpDir, "buf.db"),
      autoInstrument: [],
      storage: "local",
      trackHttp: false,
    });

    vi.spyOn(cgroup, "readMemoryPeak").mockReturnValue(256 * 1024 * 1024);

    const tt = tracker.startTask({ taskType: "lambda" });

    const handler = wrapLambdaHandler(async (_event: unknown, _ctx: unknown) => {
      return { statusCode: 200 };
    });

    const result = await runWithTask(tt.task, () => handler({}, {}));
    expect(result).toEqual({ statusCode: 200 });

    const events = tracker.buffer.queryEvents(tt.task.taskId);
    const compute = events.filter((e) => e.eventType === "compute_cost");
    expect(compute.length).toBe(1);
    const details = compute[0].details as Record<string, any>;
    expect(details.billing_model).toBe("lambda");
    expect(details.invocation_count).toBe(1);
    // 1024 MB → DECIMAL bytes (AWS env var convention).
    expect(details.memory_bytes_limit).toBe(1024 * 1_000_000);
    expect(details.memory_bytes_peak).toBe(256 * 1024 * 1024);
    expect(details.initialization_type).toBe("on-demand");
    expect(details.region).toBe("us-east-1");
    expect(["x86_64", "arm64"]).toContain(details.architecture);
    expect(details.cost_pending).toBe(true);
    expect(compute[0].costUsd.toNumber()).toBe(0); // back-fills at task finalize
  });

  test("no active task → pass-through (capture spec §6 case 2)", async () => {
    const handler = wrapLambdaHandler(async () => "ok");
    const result = await handler({}, {});
    expect(result).toBe("ok");
  });

  test("handler exception still emits event (capture spec §6 case 7)", async () => {
    process.env.AWS_LAMBDA_FUNCTION_NAME = "fn";
    process.env.AWS_LAMBDA_FUNCTION_MEMORY_SIZE = "512";
    process.env.AWS_REGION = "us-east-1";

    const tracker = dexcost.init({
      dbPath: join(tmpDir, "buf.db"),
      autoInstrument: [],
      storage: "local",
      trackHttp: false,
    });

    vi.spyOn(cgroup, "readMemoryPeak").mockReturnValue(0);

    const tt = tracker.startTask({ taskType: "lambda" });

    const handler = wrapLambdaHandler(async (_event: unknown, _ctx: unknown) => {
      throw new Error("simulated handler failure");
    });

    await expect(
      runWithTask(tt.task, () => handler({}, {})),
    ).rejects.toThrow(/simulated/);

    const events = tracker.buffer.queryEvents(tt.task.taskId);
    const compute = events.filter((e) => e.eventType === "compute_cost");
    expect(compute.length).toBe(1);
    const details = compute[0].details as Record<string, any>;
    expect(details.billing_model).toBe("lambda");
  });

  test("computeBillingOverrides threaded through init()", () => {
    const tracker = dexcost.init({
      dbPath: join(tmpDir, "buf.db"),
      autoInstrument: [],
      storage: "local",
      trackHttp: false,
      computeBillingOverrides: { cloud_run: "instance" },
      k8sNodeAware: true,
    });
    expect(tracker.computeBillingOverrides).toEqual({ cloud_run: "instance" });
    expect(tracker.k8sNodeAware).toBe(true);
  });

  test("computeBillingOverrides defaults to empty", () => {
    const tracker = dexcost.init({
      dbPath: join(tmpDir, "buf.db"),
      autoInstrument: [],
      storage: "local",
      trackHttp: false,
    });
    expect(tracker.computeBillingOverrides).toEqual({});
    expect(tracker.k8sNodeAware).toBe(false);
  });

  test("all five wraps exported from top-level", async () => {
    const idx = await import("../src/index.js");
    expect(typeof idx.wrapLambdaHandler).toBe("function");
    expect(typeof idx.wrapCloudRunHandler).toBe("function");
    expect(typeof idx.wrapCloudFunctionsHandler).toBe("function");
    expect(typeof idx.wrapAzureFunctionsHandler).toBe("function");
    expect(typeof idx.wrapVercelHandler).toBe("function");
  });
});
